WebSockets Next: Allow to send authorization headers from web browsers using JavaScript clients (#45809)
quarkusio/quarkus: docs/src/main/asciidoc/websockets-next-reference.adoc

<1> The `getProduct` callback method can only be invoked if the current security identity has an `admin` role or the user is allowed to get the product detail.
<2> The error handler is invoked in case of the authorization failure.

==== Bearer token authentication

The xref:security-oidc-bearer-token-authentication.adoc[OIDC Bearer token authentication] expects that the bearer token is passed in the `Authorization` header during the initial HTTP handshake.

> Some WebSocket clients like the https://vertx.io/docs/vertx-core/java/#_websockets_on_the_client[Vert.x WebSocketClient] support configuring headers to the WebSocket opening handshake, but clients that strictly follow the https://websockets.spec.whatwg.org/#the-websocket-interface[WebSockets API] do not support headers.

Reviewer: ```suggestion
Java WebSocket clients such as <<web-sockets-next-client>>  and https://vertx.io/docs/vertx-core/java/#_websockets_on_the_client[Vert.x WebSocketClient] support adding custom headers to the WebSocket opening handshake.
However, JavaScript clients that follow the https://websockets.spec.whatwg.org/#the-websocket-interface[WebSockets API] do not support adding custom headers.
```

Suggesting referencing the WebSocket Next client as well, and starting on a new line as preferred by the Doc team

> The absence of headers proves a challenge when writing secure WebSocket client applications for a web browser with JavaScript.

Reviewer: ```suggestion
Therefore, passing a bearer access token using a custom `Authorization` header is impossible with JavaScript based WebSocket clients.
```
It is only a problem for passing the extra content, the suggested mitigations to make the upgrade secure are listed below

> The JavaScript WebSocket client only allows to configure the HTTP `Sec-WebSocket-Protocol` request header.

Reviewer: ```suggestion
The JavaScript WebSocket client only allows to configure the HTTP https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Sec-WebSocket-Protocol[Sec-WebSocket-Protocol] request header for negotiating a sub-protocol.
```

> We have seen users to take advantage of this header for purpose which it is not intended, that is to propagate the `Authorization` header.

Reviewer: ```suggestion
If absolutely necessary, the `Sec-WebSocket-Protocol` header might be used as a carrier for custom headers, to provide a workaround the https://websockets.spec.whatwg.org/#the-websocket-interface[WebSockets API] restrictions.
```

Here is an example of a JavaScript client propagating the `Authorization` header as a subprotocol value:

[source,javascript]
----
const token = getBearerToken()
const quarkusHeaderProtocol = encodeURIComponent("quarkus-header#Authorization#Bearer " + token) <1>

> const socket = new WebSocket("wss://" + location.host + "/chat/" + username, ["irrelevant", quarkusHeaderProtocol]) <2>

Reviewer: ```suggestion
const socket = new WebSocket("wss://" + location.host + "/chat/" + username, ["bearer-token-carrier", quarkusHeaderProtocol]) <2>
```

IMHO makes sense to make it a bit more technical

----
<1> Expected format for the Quarkus Header subprotocol is `quarkus-header#header-name#header-value`.
Do not forget to encode the subprotocol value as a URI component to avoid encoding issues.
<2> Indicate 2 subprotocols supported by the client, the subprotocol of your choice and the Quarkus headers subprotocol.

> For purpose of the header propagation, it is not necessary to replace the `irrelevant` subprotocol value.

Reviewer: ```suggestion
```

IMHO it is simpler to give some technical name like `bearer-token-carrier` or something similar and note that the name of this sub-protocol is irrelevant for the purpose of transporting the custom headers

Author: I think the first sentences makes it clear that users choose the protocol of their choice, so this sentence is surplus to requirements and I'll drop it.

For the WebSocket server to accept the `Authorization` passed as a sub-protocol, we must:

> * Configure our WebSocket server with the supported sub-protocols. When the WebSocket client provides a lists of a supported sub-protocols in the HTTP `Sec-WebSocket-Protocol` request header, the WebSocket server must agree to serve content with one of them.

Reviewer: ```suggestion
* Configure our WebSocket server with the supported sub-protocols. When the WebSocket client provides a lists of supported sub-protocols in the HTTP `Sec-WebSocket-Protocol` request header, the WebSocket server must agree to serve content with one of them.
```

* Enable Quarkus header subprotocol mapping to the opening WebSocket handshake request headers.

[source, properties]
----
quarkus.websockets-next.server.supported-subprotocols=irrelevant
quarkus.websockets-next.server.propagate-subprotocol-headers=true
----

[WARNING]
====

> WebSocket security model is the origin-based model and was not designed for the client-side authentication with headers or cookies.

Reviewer: ```suggestion
WebSocket security model is origin-based and is not designed for the client-side authentication with headers or cookies.
```

For example, web browsers do not enforce the Same-origin policy for the opening WebSocket handshake request.

> We strongly recommend to only enable this feature if all of following additional security measures are in place:

Reviewer: ```suggestion
When you plan to use bearer access tokens during the opening WebSocket handshake request, we strongly recommend to follow the additional security measures listed below to minimize the security risks:
```

* Restrict supported Origins to trusted Origins only with the xref:security-cors.adoc#cors-filter[CORS filter].

> * Enforce encrypted HTTP connection via TLS (use the `wss` protocol).

Reviewer: ```suggestion
* Use the `wss` protocol to enforce encrypted HTTP connection via TLS.
```

<1> The `getProduct` callback method can only be invoked if the current security identity has an `admin` role or the user is allowed to get the product detail.
<2> The error handler is invoked in case of the authorization failure.

==== Bearer token authentication

The xref:security-oidc-bearer-token-authentication.adoc[OIDC Bearer token authentication] expects that bearer tokens are passed in the `Authorization` header during the initial HTTP handshake.

> While we recommend to use WebSocket clients that support setting custom headers in the initial HTTP handshake, there are users that rely on https://websockets.spec.whatwg.org/#the-websocket-interface[WebSockets API].

Reviewer: My understanding that anyone writing JavaScript-based WS applications has no choice but to use that API, right ?

Reviewer: IMHO it makes sense to have 2 sub-sections, one for WS Client, another for WS API, right now it reads like using WS API is quite a bad idea

Author: > My understanding that anyone writing JavaScript-based WS applications has no choice but to use that API, right ?

My understanding is there are some JS clients that also implements WS and supports that. One that seem mostly trustworthy is  https://github.com/websockets/ws/issues/1925 (pointing to the issue because it is clear they are setting headers on the client side) https://github.com/websockets/ws/blob/master/lib/websocket.js#L743. The issue with this is that some users think like I would: do my employer trust this in production and why?

> IMHO it makes sense to have 2 sub-sections, one for WS Client, another for WS API, right now it reads like using WS API is quite a bad idea

I thought we need to make very clear line between what is possible and what we suggest (nothing, you are on your own). We can use to use WebSockets API, but I think it requires additional concepts like CORS and custom protocol that has message types and we would authenticate on one of these message types. If you have idea, please suggest specific test.

Reviewer: @michalvavrik IMHO it makes sense to make it clear how users who have no other option but to use WS API can approach it, as opposed to us discouraging them, with some warnings added. 
Look at Quarkus LangChain4j demos, chatbots are everywhere, opened from scripts, they can't use Quarkus Java WS Client.
We should not be recommending specific JS client libraries, as under the hood the probably do these workarounds anyway, and I agree, we don't want to take risk and guide users to use specific libraries and then be asked why did we do it...

As far as using WS API is concerned, we should give simple suggestions:
* Use sub-protocol, I'm not sure it is still the case but it was definitely done in Kubernetes as well: https://github.com/kubernetes/kubernetes/commit/714f97d7baf4975ad3aa47735a868a81a984d1f0
* Recommend custom ticketing system to mitigate the absence of the CORS control: https://github.com/quarkiverse/quarkus-langchain4j/blob/main/samples/secure-sql-chatbot/README.md#custom-websocket-ticket-system, https://github.com/quarkiverse/quarkus-langchain4j/blob/main/samples/secure-sql-chatbot/src/main/java/io/quarkiverse/langchain4j/sample/chatbot/TicketHttpUpgradeCheck.java
* Ask users to use `wss:`

Author: that makes sense @sberyozkin , I'll rewrite this later in the afternoon. Thank you

Reviewer: Thanks @michalvavrik, take your time please

Author: I completely reworked docs, let's start a new thread if you have comments. Thanks

[source,javascript]
----
const token = getBearerToken()
const socket = new WebSocket("wss://" + location.host + "/chat/" + username, ["bearer", token]); <1>
----
<1> Indicate 2 sub-protocols supported by the client, the `bearer` sub-protocol and the token sub-protocol.

The WebSocket server can also be configured with supported sub-protocols:

[source, properties]
----
quarkus.websockets-next.server.supported-subprotocols=bearer
----
<1> The `bearer` sub-protocol matches the sub-protocol option passed to the client in previous example.

In Quarkus, it is possible to map HTTP server request headers to a different headers, for example:

[source, java]
----
package io.quarkus.websockets.next.test.security;

import static io.quarkus.websockets.next.HandshakeRequest.SEC_WEBSOCKET_PROTOCOL;
import static io.vertx.core.http.HttpHeaders.AUTHORIZATION;

import jakarta.enterprise.event.Observes;

import io.vertx.ext.web.Router;

public class WebSocketTokenHeaderHandler {

    private static final String BEARER_PROTOCOL = "bearer";
    private static final String TOKEN_PREFIX = BEARER_PROTOCOL + ",";

>     void observe(@Observes Router router) {

Reviewer 2: It should be also possible to manipulate the request headers in `HttpUpgradeCheck.perform(HttpUpgradeContext)`, or is it too late for security?

Author: yes, it's too late

Reviewer 2: I see.

Author: Do you have in mind something different than this PR, like only add section: here is how to use `HttpUpgradeCheck` to re-map headers and it works if you only use security annotations and disable proactive authentication?

Author: For people who has only WebSockets Next, I suspect it wouldn't have negative effect as annotations are likely their first option and proactive auth doesn't seem like important concept here.

>         router.get().order(-250).handler(routingContext -> {

Reviewer: I wonder, would it make sense for WS-Next provide such a handler for users not having to write this code ?

Author: see https://github.com/quarkusio/quarkus/issues/42824#issuecomment-2606712091, it can be discussed

Reviewer: IMHO optionally enabling a handler which does a header conversion according the documented scheme is all right, it is not a workaround for users who have to use WS API

Author: I also think it would be easier to understand than ask users to add this handler. We must be clear in docs about security aspects, that's all.

Reviewer: Yeah, we just explain what users who really need to retain tokens can do, may be we can offer some basic support like having this handler enabled if requests, and ask users take WS API security concerns seriously

Author: Done.
